src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py

```diff
@@ -0,0 +1,326 @@
+            # reference.
+            loop_bounds = []
+            ndims = len(array_ref.children)
+            for shape in array_ref.children:
+                loop_bounds.append((shape.start.copy(), shape.stop.copy()))
+
+        # Determine the datatype of the array's values and create a
+        # scalar of that type
+        array_intrinsic = array_ref.symbol.datatype.intrinsic
+        array_precision = array_ref.symbol.datatype.precision
+        scalar_type = ScalarType(array_intrinsic, array_precision)
+
+        symbol_table = node.scope.symbol_table
+        assignment = node.ancestor(Assignment)
+
+        datatype = scalar_type
+        array_reduction = False
+        if dimension_ref and ndims > 1:
+            array_reduction = True
+            # We are reducing from one array to another
+            shape = []
+            for idx, bounds in enumerate(loop_bounds):
+                if int(dimension_literal.value)-1 == idx:
+                    pass
+                else:
+                    shape.append(bounds)
+            datatype = ArrayType(scalar_type, shape)
+
+        array_ref = node.children[0].detach()
+
+        # Create temporary variable based on the name of the intrinsic.
+        var_symbol = symbol_table.new_symbol(
+            f"{self._INTRINSIC_NAME.lower()}_var", symbol_type=DataSymbol,
+            datatype=datatype)
+        # Replace operation with a temporary variable.
+        if array_reduction:
+            array_indices = []
+            for idx in range(ndims-1):
+                array_indices.append(":")
+            reference = ArrayReference.create(var_symbol, array_indices)
+        else:
+            reference = Reference(var_symbol)
+        node.replace_with(reference)
+
+        # Create the loop iterators
+        loop_iterators = []
+        array_iterators = []
+        for idx in range(ndims):
+            loop_iterator = symbol_table.new_symbol(
+                f"i_{idx}", symbol_type=DataSymbol, datatype=INTEGER_TYPE)
+            loop_iterators.append(loop_iterator)
+            if array_reduction and idx != int(dimension_literal.value)-1:
+                array_iterators.append(loop_iterator)
+
+        # Initialise the temporary variable.
+        rhs = self._init_var(var_symbol)
+        lhs = reference.copy()
+        new_assignment = Assignment.create(lhs, rhs)
+        assignment.parent.children.insert(assignment.position, new_assignment)
+
+        array_indices = []
+        for idx in range(ndims):
+            array_indices.append(Reference(loop_iterators[idx]))
+        array_ref = ArrayReference.create(array_ref.symbol, array_indices)
+
+        statement = self._loop_body(
+            array_reduction, array_iterators, var_symbol, array_ref)
+
+        if mask_ref:
+            # A mask argument has been provided
+            for ref in mask_ref.walk(Reference):
+                if ref.name == array_ref.name:
```

Comment:
Probably there's commonality with the handling of WHERE constructs here. Please could you take a look?

Reply:
Well spotted. The rule seems to be that if the array is not indexed then it refers to the current element. I have therefore modified the code so that only References to the array (i.e. non indexed arrays) have indices added. This seems to work. I've added a test for this in `test_mask_array_indexed` in `mms_base_trans_test.py`